### PR TITLE
feat(workflow): Support multiple order_by keys in CombinedQuerysetPaginator

### PR DESCRIPTION
**Pull request by iProgramStuff**
The CombinedQuerysetPaginator is able to take multiple models, and apply pagination (combines them, sorts them, cursors them).

Great - but right now all of the passed models have to share a field to sort by. What if we want to sort by a different field on each model? 

This is a real use case. Both our alert rules and issue rules can be named, but the field for this name is different on each model ("name" and "label"). It makes sense to us and the user to sort like this, so we need to support using a different key on each model to sort.

Thinking about how to implement this and will change usages appropriately.

### CODE REVIEW

File: src/sentry/utils/cursors.py [hunk 1]
```diff
@@ -7,7 +7,7 @@
 
 class Cursor(object):
     def __init__(self, value, offset=0, is_prev=False, has_results=None):
-        self.value = int(value)
+        self.value = value
```
[the remark this answers is not in the document]
Comment (iProgramStuff) [reply src/sentry/utils/cursors.py:10]: I think that's it, that the key needs to be responsible for this. There would be some work making sure everything still works though...or maybe we need a function to convert a string value into an int and back.

[the remark this answers is not in the document]
Comment (iProgramStuff) [reply src/sentry/utils/cursors.py:10]: The paginator now casts it as appropriate which fixed the tests.

File: src/sentry/api/paginator.py [hunk 2]
```diff
@@ -512,27 +550,33 @@ def _is_asc(self, is_prev):
 
     def _build_combined_querysets(self, value, is_prev, limit, extra):
         asc = self._is_asc(is_prev)
+        combined_querysets = list()
+        for queryset in self.querysets:
+            item = queryset.first()
```
Comment (wedamija) [inline src/sentry/api/paginator.py:555]: We probably don't need to do this query if we inspect the querysets and just determine model instead.

File: src/sentry/api/paginator.py [hunk 3]
```diff
@@ -473,34 +474,71 @@ def get_result(self, limit, cursor=None):
...
+                    key == "date_added"
+                ), "When sorting by date_added, it must be the key used on all querysets"
+        for i, queryset in enumerate(self.querysets):
+            if queryset.first() is not None:
+                assert hasattr(
+                    queryset.first(), self.keys[i]
+                ), "Model of type {} does not have field {}".format(
+                    type(queryset.first()), self.keys[i]
+                )
+            self.model_key_map.update({type(queryset.first()): self.keys[i]})
```
[the remark this answers is not in the document]
Comment (iProgramStuff) [reply src/sentry/api/paginator.py:518]: In `get_item_key`, I need to somehow get the value of the field given an instance of a model. I only have the model instance and I don't know the field name to grab. so I use the `model_key_map ` dict to do some mapping here. There is probably a better way to do this..

File: src/sentry/api/paginator.py [hunk 4]
```diff
@@ -473,34 +474,71 @@ def get_result(self, limit, cursor=None):
...
+        assert len(self.keys) == len(
+            querysets
+        ), "order_by must be a string or a list of equal length to querysets"
+        if "date_added" in self.keys:
+            for key in self.keys:
+                assert (
+                    key == "date_added"
+                ), "When sorting by date_added, it must be the key used on all querysets"
+        for i, queryset in enumerate(self.querysets):
+            if queryset.first() is not None:
```
[the remark this answers is not in the document]
Comment (iProgramStuff) [reply src/sentry/api/paginator.py:512]: Didn't realize that about first(). I removed all instances of it.

File: src/sentry/api/paginator.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (wedamija) [reply src/sentry/api/paginator.py]: Why can't you just do `self.model_key_map[type(queryset.first())] = self.keys[i]` though?

[the remark this answers is not in the document]
Comment (iProgramStuff) [reply src/sentry/api/paginator.py]: Oh, misunderstood your question haha. Changed this to directly access it

**Review by wedamija — COMMENTED** [review]
I think it's going to be tricky to sort by `created_by`, which you mentioned on your other pr. We can probably do it, but might be easier to make it a separate pr.

File: src/sentry/api/paginator.py [hunk 5]
```diff
@@ -472,35 +474,87 @@ def get_result(self, limit, cursor=None):
...
-        if order_by:
-            if order_by.startswith("-"):
-                self.key, self.desc = order_by[1:], True
+    def __init__(self, intermediaries, desc=False, on_results=None):
+        self.desc = desc
+        self.intermediaries = intermediaries
+        self.on_results = on_results
+        for intermediary in self.intermediaries:
+            if intermediary.is_empty:
+                self.intermediaries.remove(intermediary)
```
[the remark this answers is not in the document]
Comment (iProgramStuff) [reply src/sentry/api/paginator.py:520]: I was wondering that too but I didn't want every use of the paginator to have to check if the queryset was empty before forming the intermediaries array and thought this felt like a more centralized place for the logic. Definitely going to fix the logic fail here though.

File: tests/sentry/api/test_paginator.py [hunk 6]
```diff
@@ -599,10 +603,114 @@ def test_simple(self):
...
+        alert_rule0 = self.create_alert_rule(name="alertrule0")
+        alert_rule1 = self.create_alert_rule(name="alertrule1")
+        assert Rule.objects.all().count() == 1
+        rule0 = Rule.objects.all().first()
+        rule1 = Rule.objects.create(label="rule1", project=self.project)
+        alert_rule2 = self.create_alert_rule(name="alertrule2")
+        alert_rule3 = self.create_alert_rule(name="alertrule3")
+        # rule2 = Rule.objects.create(label="rule2", project=self.project)
+        rule3 = Rule.objects.create(label="rule3", project=self.project)
+        # rule4 = Rule.objects.create(label="A", project=self.project)
```
Comment (wedamija) [inline tests/sentry/api/test_paginator.py:625]: Remove commented out lines

File: tests/sentry/api/test_paginator.py [hunk 7]
```diff
@@ -599,10 +603,114 @@ def test_simple(self):
...
+        assert page2_results[2].id == alert_rule1.id
+
+        next_cursor = result.next
+        result = paginator.get_result(limit=3, cursor=next_cursor)
+        page3_results = list(result)
+        assert len(result) == 3
+        # assert page3_results[0].id == alert_rule1.id
+        assert page3_results[0].id == alert_rule0.id
+        assert page3_results[1].id == rule5.id
+        assert page3_results[2].id == rule6.id
```
Comment (wedamija) [inline tests/sentry/api/test_paginator.py:660]: It's a little weird that this is at the end. I get why that is the case, but I think it'd make sense to sort in a case insensitive way.

```
from django.db.models.functions import Lower
MyModel.objects.order_by(Lower('myfield'))
```

Probably not necessary in this pr, maybe a follow-up.

File: tests/sentry/api/test_paginator.py [hunk 8]
```diff
@@ -599,10 +603,114 @@ def test_simple(self):
...
+
+        next_cursor = result.next
+        result = paginator.get_result(limit=3, cursor=next_cursor)
+        page4_results = list(result)
+        assert len(result) == 1
+        # assert page4_results[0].id == rule6.id
```
Comment (wedamija) [inline tests/sentry/api/test_paginator.py:666]: Nit: Remove

File: src/sentry/incidents/endpoints/organization_alert_rule_index.py [hunk 9]
```diff
@@ -34,13 +38,18 @@ def get(self, request, organization):
         issue_rules = Rule.objects.filter(
             status__in=[RuleStatus.ACTIVE, RuleStatus.INACTIVE], project__in=project_ids
         )
+
+        sort_key = request.GET.get("sort", "date_added")
+        rule_sort_key = sort_key if sort_key != "name" else "label"
```
Comment (wedamija) [inline src/sentry/incidents/endpoints/organization_alert_rule_index.py:43]: Could you leave a comment explaining why we do this?

**Review by wedamija — COMMENTED** [review]
I didn't mark them all, but there are a fair few commented out lines, could you remove?

Looking good, just fix that broken test and I think we'll be good to go!

**Review by wedamija — APPROVED** [review]

**Pull request merged**